Mathlib/Algebra/Operad/Basic.lean

```diff
@@ -0,0 +1,100 @@
+
+/-! This file defines the typeclasses and notations used for workign with `Operad`, `SymmOperad`,
+  and `Clone`. These all, in some form another, work with graded - or Sigma - types. Below,
+  `A : ℕ → Type*` is always the carrier.
+
+  Classes:
+  - `MultiComposable`: A `compose` operation `A n → Fin n → A m → A (n+m-1)`, for composing an
+    element of grade `m` into one of grade `n` at a location `Fin n`. This is the type of operation
+    used in Operads.
+  - `Superposable`: A `superpose` operation `A n → (Fin n → A m) → A m`. This is the type of
+    operation used in Clones.
+  - `OneGradedOne`: There is a distinguished `1` element at grade 1, a notion of identity.
+  - `SigmaMulAction`: A graded `MulAction M A`, where elements `M i` can act on `A i` at each i.
+
+  Definitions:
+  - `composeAt` bundles the action of `MultiComposable.compose` into `Sigma A`.
+  - `superpose` bundles the action of `Superposable.superpose` into `Sigma A`.
+
+  Notations:
+  - `x ∘⟨p⟩ y` for `MultiComposable.compose` to compose `x ∘ y` at index `p`.
+  - `x ∘[p] y` for the `composeAt` Sigma variant.
+  - `x ∘⚟ y` for `Superposable.superpose x y`. The typography is meant to suggest
+    "many arguments into one".
+-/
+
+/-- A MultiComposable is a structure that allows composition from an m-arity object
+ into a n-arity object at location p (in the range 0 to n-1) to produce an (n+m-1)
+ arity object. `Operad` is the canonical example. -/
+class MultiComposable  (A : ℕ → Type*) where
+  /-- Compose the (n+1)-arity object at location p on the m-arity object. -/
+  compose {n m} : A n → Fin n → A m → A (n+m-1)
+
+/-- A Superposable is a structure that allows "superposition": given an n-arity object
+ and n many m-arity objects, they can all enter and share arguments to make a new m-arity
+object. `Clone` is the canonical example. -/
+class Superposable (A : ℕ → Type*) where
+  /-- Compose the (n+1)-arity object at location p on the m-arity object. -/
+  superpose {n m} : A n → (Fin n → A m) → A m
+
+/-- Families that have a "one" at grading 1. -/
+class OneGradedOne (A : ℕ → Type*) extends One (A 1) where
+
+variable {A} {m : ℕ}
+
+/-- Upgrade `MultiComposable.compose` to an operation on Sigma types. -/
+def composeAt [MultiComposable A] (x : Sigma A) (y : Sigma A) (p : Fin x.fst) : Sigma A :=
+  ⟨_, MultiComposable.compose x.snd p y.snd⟩
+
+/-- Upgrade `Superposable.superpose` to an operation on Sigma types. -/
+def superpose [Superposable A] (x : Sigma A) (y : Fin x.fst → A m) : Sigma A :=
+  ⟨m, Superposable.superpose x.snd y⟩
+
+/-- Notation for `MultiComposable.compose`, for working with `A n` graded types. -/
+notation:70 x:71 " ∘⟨" p:70 "⟩ " y:70  => MultiComposable.compose x p y
+/-- Notation for `composeAt`, for working with Sigma types. -/
+notation:70 x:71 " ∘[" p:70 "] " y:70  => composeAt x y p
+
+/-- Notation for `Superposable.superpose`, for working with `A n` graded types. -/
+infixr:70 " ∘⚟ " => Superposable.superpose
```

Comment:
I think we should make this notation scoped, at least for now. Could you please do that?